atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2502,6 +2538,9 @@ protected ListenableFuture<Map<Long, Long>> getCommitTimestamps(
                             timestampValidationReadCache.putAlreadyCommittedTransaction(startTs, commitTs);
                         }
                     }
+                    if (!schema4Gets.isEmpty()) {
+                        throwIfTransactionsTableSweptBeyondReadOnlyTxn();
+                    }
                     return startToCommitTimestamps;
```

Comment:
i know this will be called for every read :D

Reply:
Interestingly, if we reach here and don't throw then we could cache the result. Let's discuss how this thing can be optimised with @jeremyk-91 
For now, can you extract all of this into a class so we can reason about it better?